azavea/pfb-network-connectivity
# Use Django Q to create AnalysisBatch objects asynchronously

## Overview

Since large batch creation doesn't fit in the request/response cycle, due to each creation making multiple HTTP calls to trigger jobs and push files to S3, we needed to quickly refactor the endpoint to run asynchronously.

This PR does that using [Django Q](https://django-q.readthedocs.io/en/latest/) configured to use the ORM backend, since we're using it only for a single task, which would only be run occasionally by site admins.

### Demo

**Docker console job output after submitting batch:**

![screen shot 2018-06-12 at 6 19 41 pm](https://user-images.githubusercontent.com/1818302/41320681-bcc6b9e2-6e6e-11e8-8bb0-1d1058c92d0c.png)

** Staging papertrail output after submitting batch:**
![screen shot 2018-06-12 at 6 21 00 pm](https://user-images.githubusercontent.com/1818302/41320692-cf1ca110-6e6e-11e8-823e-f6646bc62b25.png)

### Notes

Write up in #578 describes why Django Q was selected over other solutions. Turns out this actually was pretty easy, it was trivial to setup the library for such a simple job.

The DJANGO_Q settings are "sensible defaults" (or so I think) for the current usage pattern, which is occasional, one-at-time jobs on a single worker. We can tweak these separately later if we add more tasks or experience issues.

I bumped the Django 1.11 version to 1.11.13 since that's been released since we did the Django upgrade. Didn't notice any issues.

## Testing Instructions

```
./scripts/update
./scripts/server
```

Go to http://localhost:9301/#/admin/analysis-jobs/ and click the "Run Batch" button. Upload a shapefile with a good number of boundaries, such as [PFB 100 Cities](https://s3.amazonaws.com/pfb-batch-run-inputs-us-east-1/PFB_100_cities.zip). The upload should now complete in a few seconds, and redirect you back to an empty Analysis Jobs page. At the same time, you should see docker console output regarding job creation coming from the django-q container. Once this starts, if you refresh the analysis jobs list page, you'll see jobs begin to populate the list.


_Note:_ Can also test on staging. I deployed 147783485d1d1eca2698f4956638645e88fba5db to ensure the Terraform changes work as expected.

Closes  #578
